src/SFML/Audio/Music.cpp

```diff
@@ -79,10 +79,13 @@ Music& Music::operator=(Music&&) noexcept = default;
 
 
 ////////////////////////////////////////////////////////////
-std::optional<Music> Music::tryOpenFromFile(std::optional<InputSoundFile>&& optFile)
+std::optional<Music> Music::tryOpenFromInputSoundFile(std::optional<InputSoundFile>&& optFile, const char* errorContext)
```

Comment:
I'm against providing an errorContext parameter. Handle/lig the error at the level you have the information.

Also I don't think we should use char* as param

Reply:
I think that both using an `errorContext` parameter and using `const char*` are the right choices in this case. Consider what the code would look like if we did not use an `errorContext` parameter:

```cpp
////////////////////////////////////////////////////////////
std::optional<Music> Music::tryOpenFromInputSoundFile(std::optional<InputSoundFile>&& optFile, const char* errorContext)
{
    if (!optFile.has_value())
        return std::nullopt;

    return Music(std::move(*optFile));
}


////////////////////////////////////////////////////////////
std::optional<Music> Music::openFromFile(const std::filesystem::path& filename)
{
    if (auto result = tryOpenFromInputSoundFile(InputSoundFile::openFromFile(filename)))
        return result;

    err() << "Failed to open music from file" << std::endl;
    return std::nullopt;
}


////////////////////////////////////////////////////////////
std::optional<Music> Music::openFromMemory(const void* data, std::size_t sizeInBytes)
{
    if (auto result = tryOpenFromInputSoundFile(InputSoundFile::openFromMemory(data, sizeInBytes)))
        return result;

    err() << "Failed to open music from memory" << std::endl;
    return std::nullopt;
}


////////////////////////////////////////////////////////////
std::optional<Music> Music::openFromStream(InputStream& stream)
{
    if (auto result = tryOpenFromInputSoundFile(InputSoundFile::openFromStream(stream)))
        return result;

    err() << "Failed to open music from stream" << std::endl;
    return std::nullopt;
}
```

Compare that with the currently proposed change:

```cpp
////////////////////////////////////////////////////////////
std::optional<Music> Music::tryOpenFromInputSoundFile(std::optional<InputSoundFile>&& optFile, const char* errorContext)
{
    if (!optFile.has_value())
    {
        err() << "Failed to open music from " << errorContext << std::endl;
        return std::nullopt;
    }

    return Music(std::move(*optFile));
}


////////////////////////////////////////////////////////////
std::optional<Music> Music::openFromFile(const std::filesystem::path& filename)
{
    return tryOpenFromInputSoundFile(InputSoundFile::openFromFile(filename), "file");
}


////////////////////////////////////////////////////////////
std::optional<Music> Music::openFromMemory(const void* data, std::size_t sizeInBytes)
{
    return tryOpenFromInputSoundFile(InputSoundFile::openFromMemory(data, sizeInBytes), "memory");
}


////////////////////////////////////////////////////////////
std::optional<Music> Music::openFromStream(InputStream& stream)
{
    return tryOpenFromInputSoundFile(InputSoundFile::openFromStream(stream), "stream");
}
```

Not using `errorContext` would be a clear violation of the DRY principle. There are situations where having a bit of repetition is better than having one more layer of abstraction, but this is not the case because:

- We already have a DRY helper function (`tryOpenFromInputSoundFile`)
- The helper function is `private` 
- The error printing mechanism does not capture file/line information, so producing the error in the helper function does not affect the accuracy of the error reporting

`const char*` here is also the simplest type that allows us to reduce the repetition without introducing any overhead, complexity, or having to include any additional header. Since the helper function is `private` and all of its uses are right below it, we can clearly see that only string literals are passed to `errorContext`, so the use of `const char*` is appropriate.

Reply:
DRY shouldn't be blindly applied. We do have a difference in implementation, as such it's not a simple repetition, which again is indicated by the fact that we have to provide context information. If you need context information from multiple code locations, it's a strong indication that you should in fact "duplicate" the code.

In my experiencing passing around such error information is a bad design.

Anyways, feel free to just "Resolve conversation", if you don't want to adjust it. No need to drag it out into another bikeshedding "discussion".